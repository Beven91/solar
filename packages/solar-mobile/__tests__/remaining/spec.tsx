import React from 'react';
import Remaining, { CurrentState } from '../../src/remaining';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('Remaining', () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });

  it('default', () => {
    const tree = mount(
      <Remaining remaining={55} />
    );
    // 断言：开始倒计时前的UI渲染
    expect(toJson(tree)).toMatchSnapshot();
    // 时间前进3秒
    jest.advanceTimersByTime(3000);
    tree.update();
    // 断言：3秒后的UI显示
    expect(toJson(tree)).toMatchSnapshot();

    const instance = tree.instance() as any;
    const id = instance.timerId;
    // 测试销毁
    tree.unmount();
    // 断言：在销毁时，应该清除setInterval
    expect(clearInterval).toBeCalledWith(id);
  });

  it('onend', () => {
    const fn = jest.fn();
    const tree = mount(
      <Remaining remaining={0} onEnd={fn} />
    );
    // 断言：由于设置的remaining值为0 所以会立即调用onEnd
    expect(fn).toHaveBeenCalledTimes(1);
    // 断言：开始倒计时前的UI渲染
    expect(toJson(tree)).toMatchSnapshot();

    // 更新属性remaining
    tree.setProps({ remaining: 3 });

    // 时间前进4秒
    jest.advanceTimersByTime(4000);
    tree.update();
    // 断言：4秒后的UI显示
    expect(toJson(tree)).toMatchSnapshot();
    // 断言：4秒后倒计时已结束，会再次调用onEnd
    expect(fn).toHaveBeenCalledTimes(2);

    // 再次更新属性:设置onEnd 以及remaining 为1 测试在onEnd不设置情况下不出异常
    tree.setProps({ onEnd: null, remaining: 1 });
    // 时间前进2秒
    jest.advanceTimersByTime(2000);
  });


  it('custom', () => {
    function handleProcess(date: CurrentState) {
      return (
        <div className="custom">
          {date.hour}:{date.minutes}:{date.seconds}
        </div>
      );
    }
    const tree = mount(
      <div>
        <Remaining remaining={86400} onProcessing={handleProcess} />
      </div>
    );
    // 时间前进3秒
    jest.advanceTimersByTime(3000);
    tree.update();
    // 断言：3秒后的UI显示
    expect(toJson(tree)).toMatchSnapshot();
  });


  it('format {day}天 {hour}：{minutes}：{seconds}', () => {
    const tree = mount(
      <Remaining remaining={160000} format="{day}天 {hour}：{minutes}：{seconds}" unit="day" />
    );
    // 时间前进3秒
    jest.advanceTimersByTime(3000);
    tree.update();
    // 断言：3秒后的UI显示
    expect(toJson(tree)).toMatchSnapshot();
  });


  it('format hour', () => {
    const tree = mount(
      <Remaining remaining={3605} unit="hour" />
    );
    // 时间前进3秒
    jest.advanceTimersByTime(3000);
    tree.update();
    // 断言：3秒后的UI显示
    expect(toJson(tree)).toMatchSnapshot();
  });


  it('format minutes', () => {
    const tree = mount(
      <Remaining remaining={3500} unit="minutes" format="{minutes}：{seconds}" />
    );
    // 时间前进3秒
    jest.advanceTimersByTime(3000);
    tree.update();
    // 断言：3秒后的UI显示
    expect(toJson(tree)).toMatchSnapshot();
  });


  it('format seconds', () => {
    const tree = mount(
      <Remaining remaining={60} unit="seconds" format="{seconds}" />
    );
    // 时间前进3秒
    jest.advanceTimersByTime(3000);
    tree.update();
    // 断言：3秒后的UI显示
    expect(toJson(tree)).toMatchSnapshot();
  });


  it('componentWillReceiveProps', () => {
    const tree = mount(
      <Remaining remaining={60} unit="seconds" format="{seconds}" />
    );
    const remaining = tree.instance() as Remaining;
    remaining.onRun = jest.fn();

    tree.setProps({ remaining: 10 });
    // 断言：当设置的倒计时总量发生变更时，会重置
    expect(remaining.onRun).toHaveBeenCalledTimes(1);

    tree.setProps({ remaining: 10 });
    // 断言：当设置倒计时总量与之前一致时，不执行重置
    expect(remaining.onRun).toHaveBeenCalledTimes(1);
  });


  it('format function', () => {
    const remaining = new Remaining({ remaining: undefined });

    // 测试在没有传递 data以及没有设置format情况下，是否会出异常
    expect(remaining.format(null)).toBeUndefined();

    const remaining2 = new Remaining({ remaining: undefined, format: '{name}:{seconds}' });
    expect(remaining2.format({ name: '剩余时间', seconds: '10' })).toBe('剩余时间:10');
  });

  it('renderDisplay', () => {
    const tree = mount(
      <Remaining remaining={60} unit="seconds" format="{seconds}" />
    );

    const instance = tree.instance() as Remaining;
    tree.setState({ date: undefined });

    instance.renderDisplay();
  });
});

