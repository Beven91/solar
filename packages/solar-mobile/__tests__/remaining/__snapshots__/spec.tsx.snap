// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Remaining custom 1`] = `
<div>
  <Remaining
    format="{day} {hour}：{minutes}：{seconds}"
    onProcessing={[Function]}
    remaining={86400}
    unit="hour"
  >
    <div
      className="remaining processing"
    >
      <div
        className="custom"
      >
        23
        :
        59
        :
        57
      </div>
    </div>
  </Remaining>
</div>
`;

exports[`Remaining default 1`] = `
<Remaining
  format="{day} {hour}：{minutes}：{seconds}"
  remaining={55}
  unit="hour"
>
  <div
    className="remaining processing"
  />
</Remaining>
`;

exports[`Remaining default 2`] = `
<Remaining
  format="{day} {hour}：{minutes}：{seconds}"
  remaining={55}
  unit="hour"
>
  <div
    className="remaining processing"
  >
     00：00：52
  </div>
</Remaining>
`;

exports[`Remaining format {day}天 {hour}：{minutes}：{seconds} 1`] = `
<Remaining
  format="{day}天 {hour}：{minutes}：{seconds}"
  remaining={160000}
  unit="day"
>
  <div
    className="remaining processing"
  >
    01天 20：26：37
  </div>
</Remaining>
`;

exports[`Remaining format hour 1`] = `
<Remaining
  format="{day} {hour}：{minutes}：{seconds}"
  remaining={3605}
  unit="hour"
>
  <div
    className="remaining processing"
  >
     01：00：02
  </div>
</Remaining>
`;

exports[`Remaining format minutes 1`] = `
<Remaining
  format="{minutes}：{seconds}"
  remaining={3500}
  unit="minutes"
>
  <div
    className="remaining processing"
  >
    58：17
  </div>
</Remaining>
`;

exports[`Remaining format seconds 1`] = `
<Remaining
  format="{seconds}"
  remaining={60}
  unit="seconds"
>
  <div
    className="remaining processing"
  >
    57
  </div>
</Remaining>
`;

exports[`Remaining onend 1`] = `
<Remaining
  format="{day} {hour}：{minutes}：{seconds}"
  onEnd={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  remaining={0}
  unit="hour"
/>
`;

exports[`Remaining onend 2`] = `
<Remaining
  format="{day} {hour}：{minutes}：{seconds}"
  onEnd={
    [MockFunction] {
      "calls": Array [
        Array [],
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  remaining={3}
  unit="hour"
/>
`;
